Add `dotnetup` `preview` `release-strategy-plan.md` (#54735)
dotnet/sdk: documentation/general/dotnetup/designs/dotnetup-release-strategy.md

  (`x-azure-ref` present, `X-Cache: TCP_MISS`, `Cache-Control: public, max-age=604800` = 7 days,
  `x-ms-blob-type: BlockBlob`). dotnetup daily already lands here today.
- `builds.dotnet.microsoft.com` fronts `dotnetcli.blob.core.windows.net` via **Akamai** (`Akamai-GRN` present, no `x-azure-ref`). Reaching it requires a promotion (copy) from `dotnetbuilds` → `dotnetcli`, gated by the release team's staging process. The account→CDN mapping is encoded in Arcade's `PublishingConstants.cs`.

- dotnetup's Arcade channel, with `targetFeeds: DotNetToolsFeeds`) publishes installers/checksums to `dotnetbuilds/public` and
  `dotnetbuilds/public-checksums`. i.e. we are **already in staging**.

- The versioned blob layout `dotnetup/$(DotnetupVersion)/dotnetup-$(TargetRid)$(_DotnetupNativeExt)`
  is set in [eng/Publishing.props](../../../../eng/Publishing.props) (`SetDotnetupBlobPaths`).

- The install scripts default to `Quality=daily` and resolve `https://aka.ms/dotnet/dotnetup/<quality>`
  ([scripts/get-dotnetup.sh](../../../../scripts/get-dotnetup.sh),
  [scripts/get-dotnetup.ps1](../../../../scripts/get-dotnetup.ps1)).

---

## 3. Comparisons

> - **aspire** — owns its own download site (`aspire.dev`) plus npm, WinGet, other packages, and GitHub Releases. Includes tich channel identity (stable / staging / daily / local / pr), release branches, and SHA-derived staging feeds. Promotion is *manual*: the release pipeline is queued against a selected build, extracts its BAR ID, and runs `darc add-build-to-channel`. Product SemVer is hand-authored in `eng/Versions.props`, not auto-bumped.

Author: ```suggestion
- **aspire** — owns its own download site (`aspire.dev`) plus npm, WinGet, other packages, and GitHub Releases. Includes rich channel identity (stable / staging / daily / local / pr), release branches, and SHA-derived staging feeds. Promotion is *manual*: the release pipeline is queued against a selected build, extracts its BAR ID, and runs `darc add-build-to-channel`. Product SemVer is hand-authored in `eng/Versions.props`, not auto-bumped.
```

   Archival CI URLs exist per version and are confirmed live today, e.g.
   `https://ci.dot.net/public/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe` and its
   `.../public-checksums/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe.sha512` sidecar.
   **Open:** the blob retention/TTL on `dotnetbuilds` is not yet confirmed — if it is finite,
   rollback past that window would require re-publishing rather than just repointing.

**Recovery — fix needed (e.g. security patch on an old branch):**
1. Check out the tagged commit.
2. Create `release/dnup/<tag-version>/hotfix-x.y.z` off the tag.
3. Open an internal PR with the fix into that branch and get approval.
4. Run the release pipeline (which runs tests) and repoint the selector.

This phase allows a public preview.

In this phase, before public preview we'd remove the fallback to the .NET Install Script, but keep the fallback on the `daily` dotnetup builds, and keep `release/dnup` using the `daily` dotnetup builds to build. (This prevents a broken `dotnetup` from preventing us from shipping a new `dotnetup`.)

### Phase 2 — Stable on `builds.dotnet.microsoft.com` (tentative)

> Host a `stable` URL on `builds.dotnet.microsoft.com` by coordinating with the release team to use their promotion pipeline — we already push to `dotnetbuilds`, so this is a copy/promote from

Author: ```suggestion
Migrate to `stable` versioning as well. Confirm whether it is actually feasible and correct to only support a latest version if it becomes a 'released' product.

Host a `stable` URL on `builds.dotnet.microsoft.com` by coordinating with the release team to use their promotion pipeline — we already push to `dotnetbuilds`, so this is a copy/promote from
```

- `https://ci.dot.net/public/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe`
- `https://ci.dot.net/public-checksums/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe.sha512`

---

## 5. Plan (phased)

### Phase 0 — Current

Daily channel available. `aka.ms/dotnet/dotnetup/daily/...` points at the latest published daily
build on `ci.dot.net`. SDK `main` consumes daily. No tags, no GitHub Release, no blessed selector.
The dotnetup CI pipeline ([.vsts-dnup-ci.yml](../../../../.vsts-dnup-ci.yml)) triggers on the
`dnup`, `release/dnup`, and `release/dotnetup` branches today.

### Phase 1 — Blessed `preview` selector

1. **Maintain one branch: `release/dnup`.**
2. **Expose both `/daily/` and `/preview/` URLs.** Daily keeps tracking the moving tip; preview resolves to a 'blessed' build. (But still not 'blessed' to actually tell external customers to use in production.)
3. **Change which `/daily/` build `/preview/` points at via an explicit process**
   Implement a `release` pipeline that lets an operator select a set of dotnetup release artifacts from a prior daily pipeline run.
4. **The release pipeline:**

>    - Bumps the patch version using global `msbuild` parameters and determines a preview tag.

Author: ```suggestion
   - Bumps the patch version by 1 off of the last tagged preview patch version using global `msbuild` parameters. Coordinate to push a PR via maestro or via actions that bumps the version in the real branch. Produce the preview versioned dotnetup `tag`.
```

>    - Bumps the patch version by 1 off of the last tagged preview patch version using global `msbuild` parameters. Coordinate to push a PR via maestro or via actions that bumps the version in the real branch. Produce the preview versioned dotnetup `tag`.

Author: I recognize this happens via a vendor in the SDK and we could utilize that strategy instead of trying to be fancy.  It does make me very sad we need PRs to bump the patch number.

   - Pushes that tag onto the commit from the selected daily pipeline run.

>    - Sets `preview` version metadata property as a global property override using the same methodology as the .NET SDK.

Author: ```suggestion
   - Sets `preview` version metadata property `PreReleaseVersionLabel` as a global property override using the same methodology as the .NET SDK.
```

   - Runs tests on that branch.

>    - On pass: repins the `/preview/` `aka.ms` url (only if `test` run is not enabled), and creates the GitHub Release with change notes.

Author: This is hand-waving as to whether or not the credentials to do this automatically in CI are viable and can be more concrete once we agree on the overall plan.

**Recovery — if only a simple revert is needed:**

> 1. Re-run the `release` pipeline against an existing tag and repoint the `/preview/` URL.

Author: ```suggestion
1. Re-run the `release` pipeline against an existing tag and repoint the `/preview/` URL. Allow skipping tests if needed (break-glass.)
```

   Archival CI URLs exist per version and are confirmed live today, e.g.
   `https://ci.dot.net/public/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe` and its
   `.../public-checksums/dotnetup/0.1.4-preview.4.26303.1/dotnetup-win-x64.exe.sha512` sidecar.
   **Open:** the blob retention/TTL on `dotnetbuilds` is not yet confirmed — if it is finite,
   rollback past that window would require re-publishing rather than just repointing.

**Recovery — fix needed (e.g. security patch on an old branch):**
1. Check out the tagged commit.
2. Create `release/dnup/<tag-version>/hotfix-x.y.z` off the tag.
3. Open an internal PR with the fix into that branch and get approval.
4. Run the release pipeline (which runs tests) and repoint the selector.

> This phase allows a public preview.

Author: ```suggestion
This phase allows a public preview. I view arcade rollout as valid around this time once we see stabilization of the public preview, but also potential arcade rollout with a fallback during `preview`.
```

In this phase, before public preview we'd remove the fallback to the .NET Install Script, but keep the fallback on the `daily` dotnetup builds, and keep `release/dnup` using the `daily` dotnetup builds to build. (This prevents a broken `dotnetup` from preventing us from shipping a new `dotnetup`.)

### Phase 2 — Stable on `builds.dotnet.microsoft.com` (tentative)

Migrate to `stable` versioning as well. Confirm whether it is actually feasible and correct to only support a latest version if it becomes a 'released' product.

Host a `stable` URL on `builds.dotnet.microsoft.com` by coordinating with the release team to use their promotion pipeline — we already push to `dotnetbuilds`, so this is a copy/promote from
`dotnetbuilds` → `dotnetcli` plus storage write permissions granted by the release/dnceng team.

> Similar release process, through coordination. Possibly automate `preview` releases at this point.

Author: ```suggestion
Similar release process, through coordination. Possibly automate `preview` releases at this point. Required to rollout to any azdo task. 
```

# dotnetup `Preview` Release Strategy

`dotnetup` ships today as a small internal preview built daily from the tip of the internal CI pipeline. SDK `main` consumes that daily build.

The problem: every daily change reaches every CI consumer, so we cannot let an SDK servicing/release branch depend on a *stable* dotnetup without it being broken by an unrelated daily change once we remove the `.NET Install Script` fallback.

---

## 1. Requirements

### Decisions: preview / phase 1

> **Must have**

Reviewer: I don't know where this should go in terms of priority but I think a requirement is having a good UX around how to report issues.  Nothing is more frustrating than getting a new version of a component and discover a crash and then not know where/how to report the issue.

Author: I agree with you there, and it'd be nice if there was a 'report an issue' function that also auto-injected the build/version of `dotnetup`, although I'm not sure if this is the right document to include this.

I think this is adjacent but it'd be more clear to put this in a 'bug/triage' type of plan document than here, but for now I'll briefly mention it.

- Allow others to download prior releases, even when unsupported (the intent is to let customers roll back without our intervention). Prior versions are downloadable today as immutable per-version `ci.dot.net` URLs; the underlying `dotnetbuilds` blob retention window has no formal retention policy for these artifacts, however.

> - Easy rollback or re-release when something goes wrong.

Reviewer: I think it is useful to establish a goal in terms on time.  Easy doesn't always mean quick.  We should be able to rollback once a decision in made within x minutes.  If a patch is needed we should be able to release it within y minutes.

Author: Great point. This is a better goal to target.

`dotnetup` ships today as a small internal preview built daily from the tip of the  internal CI pipeline. SDK `main` consumes that daily build.

The problem: every daily change reaches every CI consumer, so we cannot let an SDK servicing/release branch depend on a *stable* dotnetup without it being broken by an unrelated daily change once we remove the `.NET Install Script` fallback.

---

## 1. Requirements

### Decisions: preview / phase 1

**Must have**
- Allow others to download prior releases, even when unsupported (the intent is to let customers roll back without our intervention). Prior versions are downloadable today as immutable per-version `ci.dot.net` URLs; the underlying `dotnetbuilds` blob retention window has no formal retention policy for these artifacts, however.
- Easy rollback or re-release when something goes wrong.
- Develop features without risk of them reaching a stable release.
- Preview consumption / telemetry on a dev branch to confirm stability before promoting to stable.
- Only support 'latest' preview to reduce maintenance burden and move fast.
- Compliance with MSFT policy.

**Nice to have**
- No PRs required to bump minor versions.

> - Automated changelog notes.

Author: ```suggestion
- Automated changelog notes.
-  Telemetry informed rollout success gates (this is likely not legally possible)
```

In this phase, before public preview we'd remove the fallback to the .NET Install Script, but keep the fallback on the `daily` dotnetup builds, and keep `release/dnup` using the `daily` dotnetup builds to build. (This prevents a broken `dotnetup` from preventing us from shipping a new `dotnetup`.)

### Phase 2 — Stable on `builds.dotnet.microsoft.com` (tentative)

Migrate to `stable` versioning as well. Confirm whether it is actually feasible and correct to only support a latest version if it becomes a 'released' product.

Host a `stable` URL on `builds.dotnet.microsoft.com` by coordinating with the release team to use their promotion pipeline — we already push to `dotnetbuilds`, so this is a copy/promote from
`dotnetbuilds` → `dotnetcli` plus storage write permissions granted by the release/dnceng team.

Similar release process, through coordination. Possibly automate `preview` releases at this point. Required to rollout to any azdo task. 

### Phase 3 — Package-manager acquisition (tentative)

Linux Package Manager Feed / WinGet / Homebrew / etc. Needs discussion with partners.

> ### Phase 4 — Inclusion in the .NET SDK (unplanned at this time)

Author: It might be better to remove this because I don't want to get pigeonholed into this whole other discussion topic

The `aspire` pattern (GitHub App → Actions dispatch) is more sophisticated and appropriate if we
later need cross-repo orchestration or complex release-note generation, but adds setup complexity
that isn't justified for Phase 1.

#### Minimal implementation sketch

```yaml
# In the dotnetup release pipeline (Phase 1)
- stage: GitHubRelease
  displayName: Create GitHub Release
  dependsOn: PublishPreview
  jobs:
  - job: Approval
    pool: server
    steps:
    - task: ManualValidation@1
      inputs:
        instructions: 'Approve to create GitHub Release for dotnetup $(ReleaseVersion)'
        notifyUsers: '[dotnetup team alias]'

  - job: CreateRelease
    dependsOn: Approval
    pool:
      vmImage: ubuntu-latest
    steps:
    - checkout: self
    - script: |
        gh release create "v$(ReleaseVersion)" \
          --repo dotnet/sdk \
          --title "dotnetup $(ReleaseVersion)" \
          --notes-file $(Pipeline.Workspace)/release-notes.md \
          --target $(Build.SourceVersion) \
          --draft
      env:
        GH_TOKEN: $(BotAccount-dotnet-bot-repo-PAT)
      displayName: Create draft GitHub Release

> ```

Author: ```suggestion
```

for practical timescales. The risk of silent blob deletion is low but cannot be formally ruled out
without Azure portal access from the dnceng team.

---

## Appendix B: Automated GitHub Releases — Compliance & Precedent

*Investigated 2026-06-12.*

### Summary

**Yes — CI pipelines can automatically create GitHub Releases while remaining compliant with
Microsoft policy.** Multiple dotnet/Microsoft teams already do this in production.

### Teams with automated GitHub Releases

| Team / Repo | Pattern | Auth mechanism | Human gate |
|---|---|---|---|
| **dotnet/aspire** | AzDO dispatches GitHub Actions via `aspire-repo-bot` App | GitHub App installation token | AzDO pipeline approval stage |
| **dotnet/dotnet-monitor** | AzDO pipeline runs `gh release create` | `dotnet-bot` PAT (KeyVault-backed) | `ManualValidation@1` task |
| **dotnet/android-native-tools** | AzDO `GitHubRelease@1` task | AzDO service connection | `ManualValidation@0` task |
| **cli/cli** | GitHub Actions `workflow_dispatch` | `GITHUB_TOKEN` | `production` environment approval |
| **microsoft/mcp** | AzDO job runs `gh release create` + upload | GitHub App token | Pipeline stage gate |

> ### Compliance requirements

Author: ```suggestion
```

| **dotnet/dotnet-monitor** | AzDO pipeline runs `gh release create` | `dotnet-bot` PAT (KeyVault-backed) | `ManualValidation@1` task |
| **dotnet/android-native-tools** | AzDO `GitHubRelease@1` task | AzDO service connection | `ManualValidation@0` task |
| **cli/cli** | GitHub Actions `workflow_dispatch` | `GITHUB_TOKEN` | `production` environment approval |
| **microsoft/mcp** | AzDO job runs `gh release create` + upload | GitHub App token | Pipeline stage gate |

### Compliance requirements

| Requirement | How to meet it |
|---|---|
| Signed binaries | MicroBuild signing (`MicroBuildSignType: Real`) must succeed before release stage |
| SDL / security scanning | 1ES pipeline templates auto-inject APIScan, BinSkim, CodeQL, Component Governance |
| Human accountability | `ManualValidation@1` gate or GitHub environment protection rules |
| Least-privilege auth | GitHub App installation tokens or scoped PATs in KeyVault-backed variable groups |
| Auditability | 1ES `templateContext: type: releaseJob, isProduction: true` on deployment jobs |
| Draft safety net | `isDraft: true` default so accidental gate-pass only creates a draft |

> 

Author: ```suggestion
```
